### PR TITLE
Updated derivative rules for complex QR decomposition

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Updated `qr_backward` to work correctly for complex-valued inputs.
Added `torch.qr` to list of complex tests.

The previous implementation for real-valued differentiation used equation 42 from https://arxiv.org/abs/1001.1654
The current implementation is a bit simpler but the result for the real-valued input case is the same and all tests still pass.
Derivation of complex-valued QR differentiation https://giggleliu.github.io/2019/04/02/einsumbp.html

Ref. #33152

### CODE REVIEW

**Comment by albanD** [comment]
cc @vishwakftw as the original author of the backward formula

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -2006,67 +2006,58 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
...
                                       const Tensor& R) -> Tensor {
-    // For square and deep (tall) case we refer
-    // Walter, S.F and Lehmann, L., Algorithmic Differentiation of Linear
-    // Algebra Functions with Application in Optimum Experimental Design
-    // (Extended Version) The derivative for the QR decomposition is adapted
-    // from Eq. 42 of the above reference.
-
-    // Compute R (R')^{T}
+    // References:
+    // Matthias Seeger, Asmus Hetzel, Zhenwen Dai, Eric Meissner, Neil D. Lawrence (2018). Auto-Differentiating Linear Algebra.
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:2010]: Why did you change the reference for the real case?
Also why is there a reference for the LQ decomposition?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -2006,67 +2006,58 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
...
+    // https://arxiv.org/abs/1710.08717 Section 4.3 LQ Decomposition
+    // Hai-Jun Liao, Jin-Guo Liu, Lei Wang, Tao Xiang (2019). Differentiable Programming Tensor Networks.
+    // https://arxiv.org/abs/1903.09650 Section 3. QR factorization
+    // For derivations of complex-valued input case, see https://giggleliu.github.io/2019/04/02/einsumbp.html
+
+    // Compute R conj(grad_R)^{T}
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:2016]: nit: for consistency, should we just use `^H` in all the places where we take hermitian tranposes?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -2006,67 +2006,58 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
...
                                       const Tensor& Q,
                                       const Tensor& R) -> Tensor {
-    // For square and deep (tall) case we refer
-    // Walter, S.F and Lehmann, L., Algorithmic Differentiation of Linear
-    // Algebra Functions with Application in Optimum Experimental Design
-    // (Extended Version) The derivative for the QR decomposition is adapted
-    // from Eq. 42 of the above reference.
-
-    // Compute R (R')^{T}
+    // References:
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:2009]: Could you keep the original comment that mentions that these references are only for the tall case?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -2006,67 +2006,58 @@ Tensor qr_backward(const std::vector<torch::autograd::Variable> &grads, const Te
...
         /*unitriangular=*/false);
-    Tensor grad_A;
-    if (grad_R.defined()) {
-      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1) + grad_R);
-    } else {
-      grad_A = at::matmul(Q, solve_soln_1.transpose(-2, -1));
-    }
 
-    // Successive computations involve computation of QQ^{T} which is identity when A is square
-    if (A.size(-1) != A.size(-2)) {
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:2055]: Why is this term not needed anymore?

**Review by albanD — COMMENTED** [review]
Thanks for looking into this and updating the formula!

There were discussions as well about the numerical stability of the current implementation as well as limitations for degenerate matrices. Does this new algorithm have the same limitations?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:2010]: I changed the reference because a different formula is used now for the real case as well. I couldn't get the original one to work for complex, but got this one to work both for real and complex. The current formula is simpler maybe because the steps in derivation are different from the original one (from https://arxiv.org/abs/1001.1654).
LQ decomposition is the transpose of QR decomposition, in that paper they just decided to make LQ primary for them.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:2016]: I've changed comments to use `^H`.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:2009]: Sure, done.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 4 above)
Comment (IvanYashchuk) [reply torch/csrc/autograd/FunctionsManual.cpp:2055]: Because a different formula is used where this term is not present. The terms are gathered together, I guess, resulting in fewer operations and fewer terms.

**Comment by IvanYashchuk** [comment]
> There were discussions as well about the numerical stability of the current implementation as well as limitations for degenerate matrices. Does this new algorithm have the same limitations?

Yes, I think it carries the same set of limitations as the previous implementation because the new one still doesn't use pivoting or any other techniques for improving numerical stability.

**Review by albanD — APPROVED** [review]
LGTM thanks!

**Comment by albanD** [comment]
@IvanYashchuk this has merge conflict with master because the other PR landed first and modified the exact same lines.
Could you rebase on top of master please?

**Comment by IvanYashchuk** [comment]
@albanD, done.

**Comment by albanD** [comment]
Thanks for the quick rebase! The lint is not happy with the extra space in the test file though

**Comment by IvanYashchuk** [comment]
Oh, should be fixed now.

**Comment by albanD** [comment]
Thanks!

**Comment by IvanYashchuk** [comment]
Resolved another merge conflict.

**Comment by albanD** [comment]
@IvanYashchuk sorry about this one, it keeps failing to land and then have merge conflicts...
